lib/project/pro_motion/data_table.rb

```diff
@@ -0,0 +1,187 @@
+      data_cell = cell_at(index_path)
+      table_view.deselectRowAtIndexPath(index_path, animated: true) unless data_cell[:keep_selection] == true
+      trigger_action(data_cell[:action], data_cell[:arguments], index_path) if data_cell[:action]
+    end
+
+    def tableView(_, cellForRowAtIndexPath: index_path)
+      params = index_path_to_section_index(index_path: index_path)
+      data_cell = cell_at(index_path)
+      return UITableViewCell.alloc.init unless data_cell
+      create_table_cell(data_cell)
+    end
+
+    def cell_at(index_path)
+      c = object_at_index(index_path).cell
+      set_data_cell_defaults c
+    end
+
+    def controllerWillChangeContent(controller)
+      table_view.beginUpdates
+    end
+
+    def controller(controller, didChangeObject: task, atIndexPath: index_path, forChangeType: change_type, newIndexPath: new_index_path)
+      case change_type
+      when NSFetchedResultsChangeInsert
+        table_view.insertRowsAtIndexPaths([new_index_path], withRowAnimation: UITableViewRowAnimationAutomatic)
+      when NSFetchedResultsChangeDelete
+        table_view.deleteRowsAtIndexPaths([index_path], withRowAnimation: UITableViewRowAnimationAutomatic)
+      when NSFetchedResultsChangeUpdate
+        table_view.reloadRowsAtIndexPaths([index_path], withRowAnimation: UITableViewRowAnimationAutomatic)
+      end
+    end
+
+    def controllerDidChangeContent(controller)
+      table_view.endUpdates
+    end
+
+    def fetch_controller
+      @_data ||= begin
+        data_with_scope = data_model.send(data_scope)
+        if data_with_scope.sort_descriptors.empty?
+          # Try to be smart about how we sort things if a sort descriptor doesn't exist
+          attributes = data_model.send(:attribute_names)
+          sort_attribute = nil
+          [:updated_at, :created_at, :id].each do |att|
+            sort_attribute ||= att if attributes.include?(att.to_s)
+          end
+
+          if sort_attribute
+            PM.logger.warn("The `#{data_model}` model scope `#{data_scope}` needs a sort descriptor. Add sort_by(:property) to your scope. Currently sorting by :#{sort_attribute}.")
+            data_model.send(data_scope).sort_by(sort_attribute)
+          else
+            # This is where the application says goodbye and dies in a fiery crash.
+            PM.logger.error("The `#{data_model}` model scope `#{data_scope}` needs a sort descriptor. Add sort_by(:property) to your scope.")
+          end
+        else
+          data_with_scope
+        end
+      end
+
+      @fetch_controller ||= NSFetchedResultsController.alloc.initWithFetchRequest(
+        @_data.fetch_request,
+        managedObjectContext: @_data.context,
+        sectionNameKeyPath: nil,
+        cacheName: nil
+      )
+    end
+
+    def object_at_index(i)
+      fetch_controller.objectAtIndexPath(i)
+    end
+
+    # TODO - extract these methods out from Promotion into a module perhaps?
```

Comment:
Yes, let's put it into a module. Keep it in ProMotion, but included in `PM::Table` and `PM::DataTable`.

Something like `PM::TableBuilder` ?